faucet/src/faucet.rs

```diff
@@ -170,27 +219,207 @@ async fn request_fee_assets(
-    net::server::response(&req, ())
+
+    let mut txns = Vec::new();
+    let mut errs = Vec::new();
+    for _ in 0..req.state().num_grants {
+        tracing::info!(
+            "transferring {} tokens from {} to {}",
+            req.state().grant_size,
+            net::UserAddress(faucet_addr.clone()),
+            net::UserAddress(pub_key.address())
+        );
+        let bal = wallet.balance(&AssetCode::native()).await;
+        tracing::info!("Wallet balance before transfer: {}", bal);
+        match wallet
+            .transfer(
+                Some(&faucet_addr),
+                &AssetCode::native(),
+                &[(pub_key.address(), req.state().grant_size)],
+                req.state().fee_size,
+            )
+            .await
+        {
+            Ok(receipt) => {
+                txns.push(receipt);
+            }
+            Err(err) => {
+                tracing::warn!("Failed to transfer: {}", err);
+                errs.push(err);
+            }
+        }
+    }
+
+    // Signal the record breaking thread that we have spent some records, so that it can create more
+    // by breaking up larger records. Drop our handle to the wallet (which we no longer need) so
+    // that the thread can access it.
+    drop(wallet);
+    if req
+        .state()
+        .signal_breaker_thread
+        .clone()
+        .try_send(())
+        .is_err()
+    {
+        tracing::error!("Error signalling the breaker thread. Perhaps it has crashed?");
+    }
+
+    if !txns.is_empty() {
+        // If we successfully transferred any assets, return success.
+        net::server::response(&req, txns)
+    } else {
+        // Otherwise, explain why the transactions failed.
+        let mut msgs = errs.into_iter().map(|err| format!("  - {}", err));
+        Err(faucet_server_error(FaucetError::Transfer {
+            msg: format!(
+                "All transfers failed for the following reasons:\n{}",
+                msgs.join("\n")
+            ),
+        }))
+    }
+}
+
+async fn spendable_records(
+    wallet: &CapeWallet<'static, CapeBackend<'static, LoaderMetadata>>,
+) -> impl Iterator<Item = RecordInfo> {
+    let now = wallet.lock().await.state().txn_state.validator.now();
+    wallet.records().await.filter(move |record| {
+        record.ro.asset_def.code == AssetCode::native()
+            && record.ro.freeze_flag == FreezeFlag::Unfrozen
+            && !record.on_hold(now)
+    })
+}
+
+/// Break large records into smaller records.
+///
+/// When signalled on `wakeup`, this thread will break large records into small records of size
+/// `state.grant_size`, until there are at least `state.num_records` distinct records in the wallet.
+async fn break_up_records(state: FaucetState, mut wakeup: mpsc::Receiver<()>) {
+    loop {
+        // Wait until we have few enough records that we need to break them up.
+        //
+        // This is a simulation of a condvar loop, since async condvar is unstable, hence the manual
```

Comment:
Thank you for explaining this here, I spent a couple of minutes scratching my head until I noticed the comment